### PR TITLE
Privacy reference unit tests for content blocker rules

### PR DESCRIPTION
**Pull request by ladamski**
Task/Issue URL: https://app.asana.com/0/1171671347221384/1200272854466885/f
Tech Design URL: https://app.asana.com/0/1171671347221384/1200501100360630
CC:

**Description**:
Adds reference tests for domain matching rules.  Test definitions pulled as git submodule from https://github.com/duckduckgo/privacy-reference-tests

**Steps to test this PR**:
1. update git submodules
2. cmd-U in Xcode, or run the test in `domainMatching.swift` directly 

Note CI seems to be failing but for seemingly unrelated reasons (`comscore` tests), will retry again later.

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPad

**OS Testing**:

* [ ] iOS 13
* [ ] iOS 14

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -7,3 +7,6 @@
 [submodule "submodules/privacy-grade"]
 	path = submodules/privacy-grade
 	url = https://github.com/duckduckgo/privacy-grade
+[submodule "DuckDuckGoTests/MockFiles/reference-tests"]
```
Comment (brindy) [inline .gitmodules]: Please put submodules under the submodules path

File: DuckDuckGoTests/domainMatching.swift [hunk 2]
```diff
@@ -0,0 +1,89 @@
+//
+//  domainMatching.swift
```
Comment (brindy) [inline DuckDuckGoTests/domainMatching.swift]: Please use upper case for the file name here

File: DuckDuckGoTests/domainMatching.swift [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        let trackerData = try JSONDecoder().decode(TrackerData.self, from: trackerJSON)
+        
+        let refTests = try JSONDecoder().decode(RefTests.self, from: testJSON)
+        let tests = refTests.domainTests.tests
+
+        let rules = ContentBlockerRulesBuilder(trackerData: trackerData).buildRules(withExceptions: ["duckduckgo.com"],
+                andTemporaryUnprotectedDomains: [])
+
+        for test in tests {
+            print("TEST: ", test.name)
```
Comment (brindy) [inline DuckDuckGoTests/domainMatching.swift]: Please remove or replace with os_log statement

File: DuckDuckGoTests/domainMatching.swift [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            let requestURL = URL(string: test.requestURL)
+            let siteURL = URL(string: test.siteURL)
+            let requestType = ContentBlockerRulesBuilder.resourceMapping[test.requestType]
+            let rule = rules.matchURL(url: requestURL!, topLevel: siteURL!, resourceType: requestType!)
+            let result = rule?.action
+            if test.expectAction == "block" {
```
Comment (brindy) [inline DuckDuckGoTests/domainMatching.swift]: Please use the assert equal or assert true functions
eg  `XCTAssertEquals()`

File: DuckDuckGoTests/domainMatching.swift [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        let name: String
+        let desc: String
+        let tests: [Test]
+        
+    }
+    
+    let domainTests: DomainTests
+}
+
+class DomainMatching: XCTestCase {
```
Comment (brindy) [inline DuckDuckGoTests/domainMatching.swift]: This should be called `DomainMatchingTests` and the file name should match.

**Review by brindy — CHANGES_REQUESTED** [review]
A few comments related to style / convention for you. 

The tests don’t appear to pass. Please check bitrise. Click on Checks then head to bitrise for more info. If you think they should pass ping me and I’m happy to pair on it to get it going.

File: DuckDuckGoTests/domainMatching.swift (hunk 5 above)
Comment (ladamski) [reply DuckDuckGoTests/domainMatching.swift]: Done!

File: DuckDuckGoTests/domainMatching.swift (hunk 4 above)
Comment (ladamski) [reply DuckDuckGoTests/domainMatching.swift]: Done!

File: .gitmodules (hunk 1 above)
Comment (ladamski) [reply .gitmodules]: I tried this approach first, but being a Swift nub I couldn't figure out how to map that dir to a resource bundle.  This submodule just contains JSON-based test definitions, no code.

If you have any pointers on how to do that I'd definitely appreciate it!

**Comment by ladamski** [comment]
> A few comments related to style / convention for you.
> 
> The tests don’t appear to pass. Please check bitrise. Click on Checks then head to bitrise for more info. If you think they should pass ping me and I’m happy to pair on it to get it going.

Yes as I mentioned in the intro there are 3 Comscore tests failing that seem unrelated to my tests?
```
UnitTests.EntityMappingTests
  testWhenDomainHasSubdomainThenParentEntityIsFound, XCTAssertEqual failed: ("Optional("comScore")") is not equal to ("nil")
```

The odd thing is that all tests pass on my local, and they were also passing on the previous WIP branch (https://github.com/duckduckgo/iOS/runs/2917260792) before I created this fresh (rebased) one for landing purposes.   

Lets pair up on Mon!

File: DuckDuckGoTests/DomainMatchingTests.swift [hunk 6]
```diff
@@ -0,0 +1,90 @@
+//
+//  domainMatching.swift
```
Comment (brindy) [inline DuckDuckGoTests/DomainMatchingTests.swift]: Can you update this to match the file name, please?

**Review by brindy — COMMENTED** [review]
re the submodules,  Xcode doesn't care where the files are as long as they are added to a target, and I think this is the error. 

After moving the submodule,  to add to a target, click on the UnitTests target > Build Phases > Copy Bundle Resources, then click on + and add the file.  Optionally you can also add it to the explorer view as this is just a 'virtual view' of the file system.  Just right click on a folder icon and choose "Add Files..." - generally speaking this will add it to the right target for you as well, but to check that right click on the file and choose "Show File Inspector" and make sure it's in the right target.

`GradeTests.swift` is an example of this btw, it pulls in the `privacy-grade` submodule and adds a test json file in the same way.

But yeah, happy to pair on it.

File: DuckDuckGoTests/DomainMatchingTests.swift (hunk 6 above)
Comment (ladamski) [reply DuckDuckGoTests/DomainMatchingTests.swift]: Oops sorry had missed that

File: .gitmodules (hunk 1 above)
Comment (ladamski) [reply .gitmodules]: Fixed!

**Comment by ladamski** [comment]
Ok all feedback addressed, tests green.   I squashed the last 3 commits to clean up the history

**Review by brindy — APPROVED** [review]
LGTM! :)

**Comment by brindy** [comment]
Oh one thing before you merge can you give it a more readable name for the history just to make it easier to find in the future, please. :)

**Comment by ladamski** [comment]
> Oh one thing before you merge can you give it a more readable name for the history just to make it easier to find in the future, please. :)

Done, and thank you for the prompt & thorough reviews!

**Pull request merged**